[4.0] Media fields accessibility - second try

astridx opened the pull request:
Pull Request for Issue #27571 .

Please see https://github.com/joomla/joomla-cms/pull/27650 and https://github.com/joomla/joomla-cms/pull/27712#issuecomment-580956349  for test instructions.

In this Pull Request I tried to implement a suggestion from this [comment](https://github.com/joomla/joomla-cms/pull/27650#issuecomment-579773758
) - I use the Subform Field.


Edit: 10th of February
Test Instructions - You do NOT need composer or npm

1. Create a Custom Field of type media and see, that this field did not include an alternative text in front end. I named this custom field `old`.
2. Apply this patch and create a new custom field of type media. See hat all parameters, for example directory, are working. I named this custom field `new`.
3. See, that the custom fields `new` is still shown correctly in front end. ~When you edit an article, the text field for the alternative text is not show for the `old` custom field at the begin. But when you delete the image temporarily the text field appear and you can use it.~
4. Use the new form field regardless of custom field
a) Add 
```
		<field
			name="image2"
			type="accessiblemedia"
			label="COM_CONTACT_FIELD_PARAMS_IMAGE_LABEL"
			directory="banners"
		/>

```
somewhere for example in line 276 of `/administrator/components/com_contact/forms/contact.xml`.
b) If you use my example add

		`<?php echo $this->form->renderField('image2'); ?>`

in `/administrator/components/com_contact/tmpl/contact/edit.php` line 55.
c) See that you can use the field.

astridx committed:
add param for alt

astridx committed:
change fieldtype

astridx committed:
accessiblemedia

astridx committed:
layout

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

astridx committed:
correction

brianteeman commented:
Awesome!!! - Works perfectly for me.

Two suggestions
1. Dont make it optional to add the alternative text when defining the field - you should always have one so no need for that option
2. You could instead have an option to make the alternative text a required field - not sure if that will be possible as it would only be required if you have an image

brianteeman commented:
In the generated code when the field is presented you have a small bug as the generated code is

![image](https://user-images.githubusercontent.com/1296369/73443153-12ca9d80-434e-11ea-8b63-a25a92fd660e.png)

brianteeman commented:
Sorry there is another small issue when using the field in the front end

### Before
![image](https://user-images.githubusercontent.com/1296369/73443332-60470a80-434e-11ea-8c20-412bdfac0ce0.png)

### After
![image](https://user-images.githubusercontent.com/1296369/73443387-7785f800-434e-11ea-8a74-6a247aee1d02.png)

SharkyKZ commented:
If the idea is to implement this just in media custom field, `Joomla\CMS\Form\Field\AccessiblemediaField` is not required. Otherwise it should not have a dependency on media field plugin.

SharkyKZ commented on plugins/fields/media/media.php:
```diff
@@ -38,6 +38,11 @@ public function onCustomFieldsPrepareDom($field, DOMElement $parent, Form $form)
+			$fieldNode->setAttribute('type', 'accessiblemedia');
+		}
+
 		$fieldNode->setAttribute('hide_default', 'true');
```
This has to be set on the media field.

astridx replied:
@SharkyKZ  Are you sure. If so, I don't understand it correctly. I want to overwrite this line:
https://github.com/joomla/joomla-cms/blob/a9f80d608f0c0e00abf6cdf04c1c4f9932c44c42/administrator/components/com_fields/Plugin/FieldsPlugin.php#L190

SharkyKZ replied:
I mean `hide_default` attribute.

astridx replied:
OK. That is not part of this PR. I think we should do it in a separate PR, right?

SharkyKZ replied:
It is part of this PR. `$fieldNode` is now a Subform field but `hide_default` property belongs on the media field which is now inside the subform.

Personally I'd just remove `AccessiblemediaField` and use Subfields as example. This is basic code for implementing the subform with media and alt fields:
```
// This is now the subform field.
$fieldNode->setAttribute('type', 'subform');

// Add form to subform.
$form = $fieldNode->appendChild(new DOMElement('form'));

// This is the media field.
$media = $form->appendChild(new DOMElement('field'));
$media->setAttribute('type', 'media');
$media->setAttribute('name', 'filename');
$media->setAttribute('hide_default', 'true');

// Thie is the alt field.
$alt = $form->appendChild(new DOMElement('field'));
$alt->setAttribute('type', 'text');
$alt->setAttribute('name', 'alt');
```

SharkyKZ replied:
The same is needed for other parameters like `directory`. It's currently not working.

astridx replied:
Your are right. I now understand what you mean.

astridx committed:
frontend alt text dispay correction

astridx commented:
@SharkyKZ  The idea is to use a different form field if an alternative value can be entered. Not `media`, but `accessiblemedia`. And `accessiblemedia` inherits from `subform`. 
This is not a nice object-oriented. I know. `accessiblemedia` should be a child of the media field semantically correct.

astridx commented:
> @brianteeman In the generated code when the field is presented you have a small bug as the generated code is

Thank you, done.

astridx committed:
change layout

astridx commented:
> @brianteeman Sorry there is another small issue when using the field in the front end

Thanks again. Please see if you think that's good.

Quy commented on administrator/language/en-GB/plg_fields_media.ini:
```diff
@@ -3,9 +3,13 @@
 
+PLG_FIELDS_ACCESSIBLEMEDIA_LABEL="Media"
+PLG_FIELDS_ACCESSIBLEMEDIA_PARAMS_IMAGEFILE_LABEL="Image"
+PLG_FIELDS_ACCESSIBLEMEDIA_PARAMS_ALT_TEXT_LABEL="Alternative Text for Image"	
```
```suggestion
PLG_FIELDS_ACCESSIBLEMEDIA_PARAMS_ALT_TEXT_LABEL="Alt Text"	
```

To  be consistent with the text under `Images and Links` tab.

Quy commented on plugins/fields/media/params/media.xml:
```diff
@@ -28,6 +28,18 @@
+				type="radio"
+				class="switcher"
+				default="0"
+				label="PLG_FIELDS_MEDIA_PARAMS_ALT_TEXT_LABEL"
```
Move after line 34. Should be in this order `name`, `type`, `label`.

astridx replied:
Thanks, see https://github.com/joomla/joomla-cms/pull/27712/commits/59791c809fd459dc8f0c54db2eb5640c229ae069

Quy commented on plugins/fields/media/tmpl/media.php:
```diff
@@ -23,16 +23,29 @@
+	{
+		$accessiblemedia = json_decode($imagedata, true)['accessiblemedia'];
+		$path = $accessiblemedia['imagefile'];
+		$alt = ' alt="' . $accessiblemedia['alt_text'] . '"';
```
Escape value `htmlentities`

astridx replied:
Thanks, see https://github.com/joomla/joomla-cms/pull/27712/commits/59791c809fd459dc8f0c54db2eb5640c229ae069

astridx committed:
fixes

Quy commented on administrator/language/en-GB/plg_fields_media.ini:
```diff
@@ -3,9 +3,13 @@
 PLG_FIELDS_MEDIA="Fields - Media"
 PLG_FIELDS_MEDIA_LABEL="Media (%s)"
 PLG_FIELDS_MEDIA_PARAMS_DIRECTORY_LABEL="Directory"
+PLG_FIELDS_MEDIA_PARAMS_ALT_TEXT_LABEL="Alternative Text for Image"
```
```suggestion
PLG_FIELDS_MEDIA_PARAMS_ALT_TEXT_LABEL="Alt Text"
```

astridx replied:
Done, thanks

astridx committed:
language fix

Quy commented:
No image is selected, yet it is displayed on the frontend as a broken image.

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

astridx committed:
delete param in plugin, because we want to force the use. It should not be possible to use a media field without alt text. If the user want no separate alt, he can leave the text field empty.

astridx committed:
move formsource because it belongs to the field and not to the plugin

astridx committed:
change layout

astridx committed:
make field use media variables

astridx committed:
correciton

astridx committed:
correciton

astridx committed:
correction

astridx commented:
Today I changed this PR.

I have introduced a new field named `accessiblemedia`.

In my opinion, an `alt text` belongs to a field that inserts an image.
And this is a way of making a field type available for this purpose without breaking backward compatibility. The `media` field remains. I actually wanted to expand the `media` field. That would be more semantically correct. But then I did expand the `Subfield` field. I found this easier.

I do not change anything in the custom field anymore. The `accessiblemedia` field is now automatically loaded here. And an alt text is inserted in any case. At least an empty one.

What is important to me: Now, the field can also be used in other places.

So, new test instructions:

1. Create a Custom Field of type media and see, that you have to insert an alt text and that this text is displayed in the front end. (more detailed explanation [here](https://github.com/joomla/joomla-cms/pull/27650))

2. Use the new form field
a) Add 
```
		<field
			name="image2"
			type="accessiblemedia"
			label="COM_CONTACT_FIELD_PARAMS_IMAGE_LABEL"
			hide_none="1"
		/>

```
somewhere for example in line 276 of `/administrator/components/com_contact/forms/contact.xml`.

b) If you use my example add

		`<?php echo $this->form->renderField('image2'); ?>`

in `/administrator/components/com_contact/tmpl/contact/edit.php` line 55.

c) See that you can use the field. 

For my example up to know you can not display it in the front end, because it is not saved in the database. And for this, the array value has to cast into a string. But it works. 

If you like my idea I will create a more detailed test instruction. 


I thank you for the patience with me so far and look forward to further tips for improvements.

@SharkyKZ I looked at what you think should be implemented, the addition of the text field for the alt text in the media custom field method onCustomFieldsPrepareDom. But: I liked this version better. Do you think so too? You know your way around better with Joomla, so I would reconsider if you disagreed.

@Brian It's not nice if you submit a PR and nobody cares. I am always happy that you test and you will always find something with me :)

brianteeman commented:
I will take another look tomorrow

astridx committed:
language files

astridx committed:
add parents variables

astridx committed:
Merge branch '4.0-dev' into media2

astridx committed:
https://github.com/joomla/joomla-cms/pull/27746

SharkyKZ commented:
Most code in `layouts/joomla/form/field/media/accessiblemedia.php` (except rendering part) should be in field class.

brianteeman commented:
Sorry for the delay in getting to this. @astridx could you fix the conflicts please and then I will recheck

astridx committed:
wip

astridx committed:
conflict

brianteeman commented:
Not sure if it is something I am doing wrong but after creating the field and then trying to use it to add an image to an article the article is saved BUT the image is not

astridx committed:
change code form layout to field

astridx committed:
correction

brianteeman commented:
I have tested this item :white_check_mark: successfully on 19e4ed17276af12f086e9d77a77f5a886480fd6e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27712">issues.joomla.org/tracker/joomla-cms/27712</a>.</sub>

astridx commented:
@brianteeman Thanks for your test. You are super fast. 

Now I hope that my version is technically OK.

SharkyKZ commented:
Looks better now but there are still some issues. Existing field values are not being handled.

SharkyKZ commented on libraries/src/Form/Field/AccessiblemediaField.php:
```diff
@@ -0,0 +1,234 @@
+		$xml = <<<XML
+<?xml version="1.0" encoding="utf-8"?>
+<form>
+	<fields name="accessiblemedia">
```
Field group is unnecessary.

astridx replied:
Please see https://github.com/joomla/joomla-cms/pull/27712/commits/3c8a1f4bbcffcd3bbd924fbdc4bfb25dd79f4762

SharkyKZ commented on plugins/fields/media/tmpl/media.php:
```diff
@@ -23,17 +23,33 @@
+	$alt = ' alt=""';
+	
+	// Parameter alt_text is set to YES
+	if (json_decode($imagedata, true) !== null)
```
This should be in the plugin to simplify the layout.

astridx replied:
Do you mean like here: https://github.com/joomla/joomla-cms/pull/27712/commits/572a9003dba3b2863e3e110e4d648c8aa39b0ba5

astridx replied:
I have now only left the parameters necessary for the plugin in the field. https://github.com/joomla/joomla-cms/pull/27712/commits/1f67de29b3ee8695c287b303618912597d1240dd and  https://github.com/joomla/joomla-cms/pull/27712/commits/6828c488358cd3dd456768c9c9d4803ef27df53e

SharkyKZ replied:
>Do you mean like here: 572a900

Don't need to change `value` to `apivalue`, I think. Also need to account for old (string) values. Same in `onCustomFieldsPrepareDom`.

SharkyKZ replied:
>I have now only left the parameters necessary for the plugin in the field. 1f67de2 and 6828c48

Not sure if this is correct, if the intention is to use the field outside of custom fields plugin.

astridx replied:
You are right, the intention is to use the field outside of custom fields plugin. 
The parameters that are available in the Media field and are not available in the AccessibleMedia field (authorfield, width, height, link, asset) are only used internally as far as I can see. At least these are not mentioned in the documentation. https://docs.joomla.org/Media_form_field_type

I am of the opinion that the `width` and `height` do not belong in the field, but should be set in the template. In my opinion, `link` and `author field` are not related to the image file. But maybe I'm missing something.

astridx committed:
wip

astridx committed:
wip

astridx committed:
wip

astridx committed:
wip

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

astridx committed:
old values

astridx committed:
wip

astridx committed:
wip

astridx committed:
wip

astridx commented:
@SharkyKZ Thanks for your tips.

I have corrected backwards compatibility and extended the test instructions.

1. Create a Custom Field of type media and see, that this field did not include an alternative text in front end. I named this custom field old.
2. Apply this patch and create a new custom field of type media. See hat all parameters, for example directory, are working. I named this custom field new.
3. See, that the custom field new is still shown correctly in front end. When you edit an article, the text field for the alternative text is not show at the begin. But when you delete the image temporarily the text field appear and you can use it.
4. Use the new form field
a) Add 
```
		<field
			name="image2"
			type="accessiblemedia"
			label="COM_CONTACT_FIELD_PARAMS_IMAGE_LABEL"
			directory="banners"
		/>

```
somewhere for example in line 276 of `/administrator/components/com_contact/forms/contact.xml`.
b) If you use my example add

		`<?php echo $this->form->renderField('image2'); ?>`

in `/administrator/components/com_contact/tmpl/contact/edit.php` line 55.
c) See that you can use the field.

SharkyKZ commented on plugins/fields/media/media.php:
```diff
@@ -9,11 +9,78 @@
+
+		if ($field->value === '' || json_last_error() === JSON_ERROR_NONE)
+		{
+			$fieldNode->setAttribute('type', 'accessiblemedia');
```
The field type should always be set. Existing data needs to be mapped to the new field type.

astridx replied:
If `($field->value === '' || json_last_error() === JSON_ERROR_NONE)` the type is set in the parent method.

SharkyKZ replied:
I mean type should always be `accessiblemedia`.

astridx replied:
Why do you mean type should always be accessiblemedia? 

I tried to do that. But updating the value in the methode `onCustomFieldsPrepareDom()`  is not enough. And I don't see what I have to change all.

The current implementation is flawless for me.

SharkyKZ replied:
Currently it's not possible to add alt text in articles already containing an image.

astridx replied:
@SharkyKZ ~You're right. At the moment you have to delete the image and set it new when editing the field (I wrote this in the test instructions). Then the text box appears to enter the alternative text.~

~I couldn't do it any better. I always came across other problems and found this solution to be the best. If you know a better way, you are welcome to implement it and close this.~

The field is compliant with a11y because an empty alternative text is always output.

astridx replied:
@SharkyKZ  I had a look again: Now it's not possible to add alt text in articles already containing an image

Quy commented on plugins/fields/media/tmpl/media.php:
```diff
@@ -20,20 +20,22 @@
+	if (file_exists($path))
 	{
-		continue;
+		$buffer .= sprintf('<img src="%s"%s%s>',
```
```suggestion
		$buffer = sprintf('<img src="%s"%s%s>',
```

Quy commented on plugins/fields/media/media.php:
```diff
@@ -9,11 +9,78 @@
+		}
+		else
+		{
+			return array('imagefile' => $value, 'alt_text' => '');
```
Remove outside of else section and remove empty else section.

astridx replied:
Sorry, but I don't understand what you mean.

Quy replied:
Change to:
```
		if (json_last_error() === JSON_ERROR_NONE)
		{
			return (array) json_decode($value, true);
		}
    
		return array('imagefile' => $value, 'alt_text' => '');
```

Quy commented on plugins/fields/media/media.php:
```diff
@@ -9,11 +9,78 @@
+	{
+		json_decode($value);
+
+		if (json_last_error() === JSON_ERROR_NONE) {
```
Move bracket to its own line.

astridx replied:
Done. 
(I checked with 
`./libraries/vendor/bin/phpcs --extensions=php -p --standard=libraries/vendor/joomla/cms-coding-standards/lib/Joomla-CMS .`
But this did not not show it as an coding standard error.)

SharkyKZ replied:
Better to use `joomla/coding-standards`. `joomla/cms-coding-standards` contains exclusions for CMS files currently containing CS errors.

astridx replied:
@SharkyKZ Thank you. Now I understand that with CS errors.

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

astridx committed:
cs

astridx committed:
Remove outside of else section and remove empty else section.

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

astridx committed:
nun aber

SharkyKZ commented on libraries/src/Form/Field/AccessiblemediaField.php:
```diff
@@ -0,0 +1,184 @@
+		if ($value !== '' && json_last_error() !== JSON_ERROR_NONE)
+		{
+			$value = '{"imagefile":"' . $value . '","alt_text":""}';
+		}
```
This should be in the plugin.

astridx replied:
I tried it that way. 
Then the picture of an in the past created field is only displayed correctly in the front. It is not displayed in the back end if you edit an article. 
What am I doing wrong?

SharkyKZ replied:
Not sure. But this will need to be fixed somehow :/

astridx replied:
So that the PR works again I fixed it in the field again. I see no way to influence the value of the field in the custom field - just the markup. If someone sees a possibility: I would be happy to learn. I also think that this should be done better in Custom Field.

astridx committed:
try

astridx committed:
wip

astridx commented on plugins/fields/media/media.php:
```diff
@@ -9,11 +9,78 @@
+
+		json_decode($field->value);
+
+		$field->apivalue = $this->checkValue($field->value);
```
I tried it that way.
Then the picture of an in the past created field is only displayed correctly in the front. It is not displayed in the back end if you edit an article.
What am I doing wrong?

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

astridx committed:
redo

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

brianteeman commented:
Maybe someone can convince the accessibility team to test this

astridx commented:
Should this PR get the label `Release Blocker`? I ask because the original issue #27571 had this label and is now closed.

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

zero-24 commented on plugins/fields/media/media.php:
```diff
@@ -9,11 +9,72 @@
+		// Check if the field value is an old (string) value
+		$field->apivalue = $this->checkValue($field->value);
+	}
+
```
Doc block is missing @astridx

astridx replied:
Thanks @zero-24 . Out of laziness, I often use the Joomla internal cs-checker and this doesn't show such errors.

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

astridx committed:
wip

astridx commented:
Concerning the error in the system checks:
I just run 
`libraries/vendor/bin/codecept run acceptance tests/Codeception/acceptance/administrator/components/com_media/`
on my local machine on this branch and I get no error. 

Perhaps this related to this: https://github.com/joomla/joomla-cms/pull/28660

![Screenshot from 2020 04 18 10 19 05](https://user-images.githubusercontent.com/9974686/79632328-d2193e00-815e-11ea-9d0c-c0387993a33e.png)

astridx committed:
add comment

astridx committed:
add comment

astridx committed:
cs

astridx committed:
extend check

astridx commented:
@SharkyKZ I want to make sure I understand you correctly in the comment here: https://github.com/joomla/joomla-cms/pull/27712#discussion_r379823767

In my opinion is the check you would like to have in the plugin not wrong in the field and I have now expanded it.

**Field**

If I understand you correctly, you think that this check should be done in the plugin, since it is only necessary in the plugin. It is not necessary in the Field. 

**Plugin**

In the case of the plugin it could be, that we have an "old" custom field, that was created using the `MediaField` (not the `AccessiblemediaField` - That will be the case, if Joomla Version 3 is updated to Version 4). In this case, the `value` does not contain an alt text. It is a simple string, for example: `"images/joomla_black.png"`. 
For showing everything (image and alt text) correctly in back end the `AccessiblemediaField` need da json string like this: `"{"imagefile":"images\/banners\/banner.jpg","alt_text":""}"` because it is a child of the `SubformField`.
(I did not extend the `MediaField`, because I could reuse more. And as a `SubformField` we always have the possibility to add other things that belong to the field in the same way.)

**Summary**
You're right, I created this check in order to make the value of the custom fields backwards compatible. But: A check at this point will correct other possible errors. In my opinion the check is even better at this point.

--

Do I see that correctly, or are there other reasons to check the value in the custom field that I don't see?

SharkyKZ commented:
The check should be in the plugin because it looks like it is a workaround for the plugin. Library code should not have extension specific hacks. But if the intention is to have this field as a drop-in replacement for `Joomla\CMS\Form\Field\MediaField` in other places, maybe it's fine.

astridx commented:
@SharkyKZ   Thanks for the feedback.

> The check should be in the plugin because it looks like it is a workaround for the plugin. 

I don't see any way to do this in the plugin. But I like to be instructed when others prefer it. 
In my opinion, it is even better as it is in this PR, because other possible errors are also caught here.


>  But if the intention is to have this field as a drop-in replacement for `Joomla\CMS\Form\Field\MediaField` in other places, maybe it's fine.

This field should be an alternative to the media field, where you don't have to add a text field for the alternative text in a form. Everything is together - it belongs together. 
In my opinion it is not possible to replace the previous field with backward compatibility.

richard67 commented:
@astridx Shall I resolve the conflict in file `plugins/fields/media/tmpl/media.php` for you?

astridx committed:
conflict

astridx commented:
@richard67 I think I managed it myself. But it would be good if you could check. I am not good at conflict resolution. I usually delete the branch and do it from scratch.

astridx commented:
@zero-24  I just saw that you requested a change. But I don't really know what to change. Can you help me?

richard67 commented:
@astridx Was an easy conflict and you solved it the right way. I was sure you can do that yourself, just offered my help because I saw if and thought maybe you are too busy with other things.

richard67 commented:
@astridx Possibly @zero-24 meant this comment [https://github.com/joomla/joomla-cms/pull/27712#discussion_r410661229](https://github.com/joomla/joomla-cms/pull/27712#discussion_r410661229) about the missing doc block?

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

astridx committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into media2

wilsonge commented:
I'm merging this because it's definitely a step in the right direction and we can tweak this moving forwards as required. Thankyou very much @astridx

Scrabble96 commented:
This option to add 'Alt text' needs to be added to the "Image List" field as well:
![image-list](https://user-images.githubusercontent.com/36987494/84784846-ad1c3c80-afe2-11ea-932e-8e7b18026cd8.jpg)

richard67 commented:
@Scrabble96 Could you open a new issue for that it there is not already one found by a quick view? That would be helpful. Thanks in advance.

Scrabble96 commented:
I opened the new issue (#29640) as requested, but no-one except me has commented and adding/editing any code for a PR is way beyond me. Thanks.
